full-service/src/service/gift_code.rs

```diff
@@ -0,0 +1,644 @@
+            }
+            Err(e) => return Err(e.into()),
+        };
+
+        // Note that we now need to allow the sync thread to catch up for this TXO so that we can
+        // make sure the subaddress is assigned, rendering the Txo spendable.
+        std::thread::sleep(std::time::Duration::from_secs(5));
+
+        // Construct a transaction from the gift code account to our desired recipient account.
+        let destination_address = assigned_subaddress_b58.unwrap_or_else(|| {
+            let json_address = self
+                .create_assigned_subaddress(&account_id_hex, Some(&format!("Gift Code: {}", memo)))
+                .unwrap();
+            json_address.public_address
+        });
+        log::info!(
+            self.logger,
+            "Consuming gift code to destination address {:?}",
+            destination_address
+        );
+
+        if value < MINIMUM_FEE {
+            return Err(GiftCodeServiceError::InsufficientValueForFee(value).into());
+        }
+
+        // Sanity check that our txo is available
+        let txos = Txo::list_for_account(
+            &gift_code_account_id_hex,
+            &self.wallet_db.get_password_hash()?,
+            &self.wallet_db.get_conn()?,
+        )?;
+        if txos.len() != 1 {
+            return Err(
+                GiftCodeServiceError::UnexpectedNumTxosInGiftCodeAccount(txos.len()).into(),
+            );
+        }
+        if txos[0].txo.value as u64 != value {
+            return Err(GiftCodeServiceError::UnexpectedValueInGiftCodeTxo(
+                txos[0].txo.value as u64,
+            )
+            .into());
+        }
+
+        log::info!(self.logger, "\x1b[1;33m GOT TXOS = {:?}\x1b[0m", txos);
+
+        // We go with all the defaults because there is only one TXO in this account to spend.
+        let submit_response = self.send_transaction(
+            &gift_code_account_id_hex,
+            &destination_address,
+            (MINIMUM_FEE).to_string(),
+            None,
+            None,
+            None,
+            None,
+            Some(format!("Consume Gift Code: {}", memo)),
+        )?;
+        log::info!(
+            self.logger,
+            "Submitting transaction to consume gift code with id {:?}",
+            submit_response.transaction_id
+        );
+
+        let transaction_log =
+            TransactionLog::get(&submit_response.transaction_id, &self.wallet_db.get_conn()?)?;
+
+        // Add the consumed gift code to our gift code store. If we also own this gift code, then update.
+        match GiftCode::get(&gift_code_b58, &self.wallet_db.get_conn()?) {
+            Ok(gc) => {
+                log::info!(self.logger, "Updating existing gift code consume_log_id");
+                gc.update_consume_log_id(transaction_log.id, &self.wallet_db.get_conn()?)?
+            }
+            Err(WalletDbError::GiftCode(GiftCodeDbError::GiftCodeNotFound(_))) => {
+                log::info!(self.logger, "Registering gift code");
```

Comment:
This should have more information like the b58 or the id or anything that makes it easier to debug, this applies to other log messages in this method.